source/uwp/Renderer/lib/ActionHelpers.cpp

```diff
@@ -0,0 +1,1014 @@
+using namespace ABI::AdaptiveNamespace;
+using namespace ABI::Windows::Foundation;
+using namespace ABI::Windows::Foundation::Collections;
+using namespace ABI::Windows::UI::Xaml;
+using namespace ABI::Windows::UI::Xaml::Automation;
+using namespace ABI::Windows::UI::Xaml::Controls;
+using namespace ABI::Windows::UI::Xaml::Controls::Primitives;
+using namespace ABI::Windows::UI::Xaml::Input;
+using namespace ABI::Windows::UI::Xaml::Media;
+
+namespace AdaptiveNamespace::ActionHelpers
+{
+    HRESULT GetButtonMargin(_In_ IAdaptiveActionsConfig* actionsConfig, Thickness& buttonMargin) noexcept
+    {
+        buttonMargin = {0, 0, 0, 0};
+        UINT32 buttonSpacing;
+        RETURN_IF_FAILED(actionsConfig->get_ButtonSpacing(&buttonSpacing));
+
+        ABI::AdaptiveNamespace::ActionsOrientation actionsOrientation;
+        RETURN_IF_FAILED(actionsConfig->get_ActionsOrientation(&actionsOrientation));
+
+        if (actionsOrientation == ABI::AdaptiveNamespace::ActionsOrientation::Horizontal)
+        {
+            buttonMargin.Left = buttonMargin.Right = buttonSpacing / 2;
+        }
+        else
+        {
+            buttonMargin.Top = buttonMargin.Bottom = buttonSpacing / 2;
+        }
+
+        return S_OK;
+    }
+
+    void ArrangeButtonContent(_In_ IAdaptiveActionElement* action,
+                              _In_ IAdaptiveActionsConfig* actionsConfig,
+                              _In_ IAdaptiveRenderContext* renderContext,
+                              ABI::AdaptiveNamespace::ContainerStyle containerStyle,
+                              _In_ ABI::AdaptiveNamespace::IAdaptiveHostConfig* hostConfig,
+                              bool allActionsHaveIcons,
+                              _In_ IButton* button)
+    {
+        HString title;
+        THROW_IF_FAILED(action->get_Title(title.GetAddressOf()));
+
+        HString iconUrl;
+        THROW_IF_FAILED(action->get_IconUrl(iconUrl.GetAddressOf()));
+
+        ComPtr<IButton> localButton(button);
+
+        // Check if the button has an iconUrl
+        if (iconUrl != nullptr)
+        {
+            // Get icon configs
+            ABI::AdaptiveNamespace::IconPlacement iconPlacement;
+            UINT32 iconSize;
+
+            THROW_IF_FAILED(actionsConfig->get_IconPlacement(&iconPlacement));
+            THROW_IF_FAILED(actionsConfig->get_IconSize(&iconSize));
+
+            // Define the alignment for the button contents
+            ComPtr<IStackPanel> buttonContentsStackPanel =
+                XamlHelpers::CreateXamlClass<IStackPanel>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_StackPanel));
+
+            // Create image and add it to the button
+            ComPtr<IAdaptiveImage> adaptiveImage;
+            THROW_IF_FAILED(MakeAndInitialize<AdaptiveImage>(&adaptiveImage));
+
+            adaptiveImage->put_Url(iconUrl.Get());
+            adaptiveImage->put_HorizontalAlignment(ABI::AdaptiveNamespace::HAlignment::Center);
```

Comment:
I realize this is copy paste, but since this line changed with the HString fix I now notice we're missing error checks on these calls.